File: Templates/_BusinessApp_/_BusinessApp_/Managers/UserManager.m
//
//  UserManager.m
//  MyInsurrance
//
//  Created by Gytenis Mikulėnas on 4/19/14.
//  Copyright (c) 2015 Gytenis Mikulėnas
//  https://github.com/GitTennis/SuccessFramework
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE. All rights reserved.
//

#import "UserManager.h"
#import "UserObject.h"
#import "SettingsManager.h"
#import "GMObserverList.h"
#import "UserManagerObserver.h"
#import "KeychainManager.h"

// User persistence keys
#define kUserLoggedInPersistenceKey @"LoggedInUser"
#define kUserKeychainTokenKey @"KeychainUserToken"

@interface UserManager () {
    
    GMObserverList *_observers;
}

@end

@implementation UserManager

@synthesize user = _user;
@synthesize networkOperationFactory = _networkOperationFactory;
@synthesize settingsManager = _settingsManager;
@synthesize analyticsManager = _analyticsManager;
@synthesize keychainManager = _keychainManager;

#pragma mark - Public -

#pragma mark UserManagerProtocol

- (instancetype)init {
    
    self = [super init];
    if (self) {
        
        _user = [[UserObject alloc] init];
        _observers = [[GMObserverList alloc] initWithObservedSubject:self];
    }
    return self;
}

#pragma mark User handling

- (instancetype)initWithSettingsManager:(id <SettingsManagerProtocol>)settingsManager networkOperationFactory:(id <NetworkOperationFactoryProtocol>)networkOperationFactory analyticsManager:(id<AnalyticsManagerProtocol>)analyticsManager keychainManager:(id<KeychainManagerProtocol>)keychainManager {
    
    self = [self init];
    if (self) {
        
        _settingsManager = settingsManager;
        _networkOperationFactory = networkOperationFactory;
        _analyticsManager = analyticsManager;
        _keychainManager = keychainManager;
        
        [self loadUser];
    }
    
    return self;
}

- (BOOL)isUserLoggedIn {
    
    return _user.token.length > 0;
}

- (NSError *)saveUser:(UserObject *)user {
    
    // Store token
    NSError *error = [_keychainManager setAuthentificationToken:user.token];
    
    if (!error) {
        
        // Serialize and store user in settings
        NSDictionary *dict = [user toDict];
        [_settingsManager setLoggedInUser:dict];
        
        // Update in-memory user
        _user = user;
    }
    
    return error;
}

- (void)loadUser {
    
    NSDictionary *dict = [_settingsManager loggedInUser];
    
    if (!dict) {
        
        _user = nil;
        return;
    }
    
    _user = [[UserObject alloc] initWithDict:dict];
    
    NSString *token = [_keychainManager authentificationToken];
    
    if (token) {
        
        _user.token = token;
        
    } else {
        
        // Clear user if token was not loaded
        _user = nil;
    }
}

- (void)logout {
    
    NSError *error = [_keychainManager setAuthentificationToken:nil];
    
    if (!error) {
        
        [_settingsManager setLoggedInUser:nil];
        
        _user = nil;
        
        // There should be a webservice which does login and invalides token (if somebody sniffed and stole it). However there's no such webservice and we do logout locally only and so always consider logout as success immediatelly
        [self notifyObserversWithLogoutSuccess:nil];
        
    } else {
        
        [self notifyObserversWithLogoutFail:@"Unable to clear keychain"];
    }
}

- (void)loginUserWithData:(UserObject *)data callback:(Callback)callback {
    
    /*__weak typeof(self) weakSelf = self;
    
    Callback wrappedCallback = ^(BOOL success, id result, NSError *error) {
        
        if (success) {
            
            NSError *err = [weakSelf saveUser:result];
            if (!err) {
                
                [self notifyObserversWithLoginSuccess:weakSelf.user];
                
            } else {
                
                [self notifyObserversWithLoginFail:err.localizedDescription];
            }
            
        } else {
            
            [self notifyObserversWithLoginFail:[error localizedDescription]];
        }
        
        callback(success, result, error);
    };
    
    id<NetworkOperationProtocol> userLoginOperation = [_networkOperationFactory userLoginNetworkOperationWithParams:[data toDict]];
    [userLoginOperation performWithCallback:wrappedCallback];*/
    
#warning demo code
    UserObject *user = [[UserObject alloc] init];
    user.email = @"gytenis@test.com";
    user.token = @"123";
    [self saveUser:user];
    [self notifyObserversWithLoginSuccess:user];
    callback(YES, user, nil);
}

- (void)signUpUserWithData:(UserObject *)data callback:(Callback)callback {
    
    /*__weak typeof(self) weakSelf = self;
    
    Callback wrappedCallback = ^(BOOL success, id result, NSError *error) {
        
        if (success) {
            
            NSError *err = [weakSelf saveUser:result];
            if (!err) {
                
                [self notifyObserversWithSignUpSuccess:weakSelf.user];
                
            } else {
                
                [self notifyObserversWithSignUpFail:err.localizedDescription];
            }
            
        } else {
            
            [self notifyObserversWithSignUpFail:[error localizedDescription]];
        }
        
        callback(success, result, error);
    };
    
    id<NetworkOperationProtocol> userSignUpOperation = [_networkOperationFactory userSignUpNetworkOperationWithParams:[data toDict]];
    [userSignUpOperation performWithCallback:wrappedCallback];*/
    
#warning demo code
    data.token = @"123";
    [self saveUser:data];
    [self notifyObserversWithSignUpSuccess:data];
    callback(YES, data, nil);
}

- (void)resetPassword:(UserObject *)data callback:(Callback)callback {
    
    Callback wrappedCallback = ^(BOOL success, id result, NSError *error) {
        
        if (success) {
            
            [self notifyObserversWithPasswordResetSuccess:nil];
            
        } else {
            
            [self notifyObserversWithPasswordResetFail:nil];
        }
        
        callback(success, result, error);
    };
    
    id<NetworkOperationProtocol> userResetPasswordOperation = [_networkOperationFactory userResetPasswordNetworkOperationWithParams:[data toDict]];
    [userResetPasswordOperation performWithCallback:wrappedCallback];
}

- (void)getUserProfileWithCallback:(Callback)callback {
    
    __weak typeof(self) weakSelf = self;
    
    Callback wrappedCallback = ^(BOOL success, id result, NSError *error) {
        
        if (success) {
            
            weakSelf.user = (UserObject *)result;
        }
        
        callback(success, result, error);
    };
    
    id<NetworkOperationProtocol> userProfileOperation = [_networkOperationFactory userProfileNetworkOperationWithParams:nil];
    [userProfileOperation performWithCallback:wrappedCallback];
}

#pragma mark User state observer handling

- (void)addServiceObserver:(id<UserManagerObserver>)observer {
    
    [_observers addObserver:observer];
}

- (void)removeServiceObserver:(id<UserManagerObserver>)observer {
    
    [_observers removeObserver:observer];
}

#pragma mark - Private -

#pragma mark UserManagerObserver handling

- (void)notifyObserversWithLoginSuccess:(UserObject *)userObject {
    
    [_observers notifyObserversForSelector:@selector(didLoginUser:) withObject:userObject];
}

- (void)notifyObserversWithLoginFail:(NSString *)errorMessage {
    
    [_observers notifyObserversForSelector:@selector(didFailedToLoginUser:) withObject:errorMessage];
}

- (void)notifyObserversWithSignUpSuccess:(UserObject *)userObject {
    
    [_observers notifyObserversForSelector:@selector(didSignUpUser:) withObject:userObject];
}

- (void)notifyObserversWithSignUpFail:(NSString *)errorMessage {
    
    [_observers notifyObserversForSelector:@selector(didFailedToSignUpUser:) withObject:errorMessage];
}

- (void)notifyObserversWithUpdateSuccess:(UserObject *)userObject {
    
    [_observers notifyObserversForSelector:@selector(didUpdateUser:) withObject:userObject];
}

- (void)notifyObserversWithUpdateFail:(NSString *)errorMessage {
    
    [_observers notifyObserversForSelector:@selector(didFailedToUpdateUser:) withObject:errorMessage];
}

- (void)notifyObserversWithLogoutSuccess:(UserObject *)userObject {
    
    [_observers notifyObserversForSelector:@selector(didLogoutUser:) withObject:nil];
}

- (void)notifyObserversWithLogoutFail:(NSString *)errorMessage {
    
    [_observers notifyObserversForSelector:@selector(didFailedToLogoutUser:) withObject:errorMessage];
}

- (void)notifyObserversWithPasswordResetSuccess:(NSString *)email {
    
    [_observers notifyObserversForSelector:@selector(didResetPassword:) withObject:email];
}

- (void)notifyObserversWithPasswordResetFail:(NSString *)errorMessage {
    
    [_observers notifyObserversForSelector:@selector(didFailedToResetPassword:) withObject:errorMessage];
}

@end

File: Templates/_BusinessAppSwift_/_BusinessAppSwift_/Managers/UserManager.swift
//
//  UserManager.swift
//  _BusinessAppSwift_
//
//  Created by Gytenis Mikulenas on 06/09/16.
//  Copyright © 2016 Gytenis Mikulėnas 
//  https://github.com/GitTennis/SuccessFramework
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE. All rights reserved.
//

import UIKit

class UserManager : UserManagerProtocol {

    // MARK: UserManagerProtocol     
    var settingsManager: SettingsManagerProtocol!
    var networkOperationFactory: NetworkOperationFactoryProtocol!
    var analyticsManager: AnalyticsManagerProtocol!
    var keychainManager: KeychainManagerProtocol!
    
    // Data object
    var user: UserEntity?
    
    // https://objectpartners.com/2016/01/18/exploring-swift-initializers/
    // Every class needs to have at least one designated initializer where all non optional properties are initialized
    // Such init would allow you to create empty object but then all the properties would need to be optionals
    /*required init() {
        
        // ...
    }*/
    
    // User handling
    required init(settingsManager: SettingsManagerProtocol, networkOperationFactory: NetworkOperationFactoryProtocol, analyticsManager: AnalyticsManagerProtocol, keychainManager: KeychainManagerProtocol) {
        
        self.settingsManager = settingsManager
        self.networkOperationFactory = networkOperationFactory
        self.analyticsManager = analyticsManager
        self.keychainManager = keychainManager
        
        // Load user
        let dict: Dictionary <String, AnyObject>? = settingsManager.loggedInUser
        
        if (dict == nil) {
            
            user = nil;
        }
        
        let token: String? = keychainManager.authentificationToken()
        
        if let token = token {
            
            if let dict = dict {
                
                var dictWithToken: Dictionary<String, AnyObject> = Dictionary()
                
                dictWithToken = dictWithToken.merged(with: dict)
                dictWithToken[kUserTokenKey] = token as AnyObject
                
                user = UserEntity.init(dict: dictWithToken)
            }
            
        } else {
            
            // Clear user if token was not loaded
            user = nil
        }
    }
    
    func isUserLoggedIn()->Bool {

        let intCount = user?.token?.characters.count
        
        if let intCount = intCount {
            
            return (intCount > 0)
            
        } else {
            
            return false
        }
    }
    
    func login(user: UserEntity, callback: @escaping Callback) {
        
        //weak var weakSelf = self
        //__weak typeof(wself) weakSelf = self;
        
        // Weak and unowned self:
        // https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/AutomaticReferenceCounting.html
        let wrappedCallback: Callback = { [weak self] (_ success: Bool, _ result: Any?, _ context: Any?, _ error: ErrorEntity?) -> Void in
            
            if (success) {
                
                let error: ErrorEntity? = self?.save(user: result as! UserEntity)
                
                if let error = error {
                    
                    self?.notifyObserversWithLoginFail(error: error)
                    
                } else {
                    
                    // [weakSelf.pushNotificationManager loginUser:data callback:^(BOOL success, id result, NSError *error) {
                    // ...
                    // }];
                    
                    self?.notifyObserversWithLoginSuccess(user: (self?.user!)!)
                }
                
            } else {
                
                self?.notifyObserversWithLoginFail(error: error)
            }
            
            callback(success, result, nil, error);
        }
        
        let hashedPwd: String = (user.password as NSString).sha512(withSalt: nil);
        user.password = hashedPwd
        
        let userLoginOperation: NetworkOperationProtocol = networkOperationFactory.userLoginNetworkOperation(context: user.serializedDict())
        userLoginOperation.perform(callBack: wrappedCallback)
    }
    
    func signUp(user: UserEntity, callback: @escaping Callback) {
        
        let wrappedCallback: Callback = { [weak self] (_ success: Bool, _ result: Any?, _ context: Any?, _ error: ErrorEntity?) -> Void in
            
            if (success) {
                
                let error: ErrorEntity? = self?.save(user: result as! UserEntity)
                
                if let error = error {
                    
                    self?.notifyObserversWithSignUpFail(error: error)
                    
                } else {
                    
                    // [weakSelf.pushNotificationManager signUpUser:data callback:^(BOOL success, id result, NSError *error) {
                    // ...
                    // }];
                    
                    self?.notifyObserversWithSignUpSuccess(user: (self?.user!)!)
                }
                
            } else {
                
                self?.notifyObserversWithSignUpFail(error: error)
            }
            
            callback(success, result, nil, error);
        }
        
        let hashedPwd: String = (user.password as NSString).sha512(withSalt: nil);
        user.password = hashedPwd
        
        let userSignUpOperation: NetworkOperationProtocol = networkOperationFactory.userSignUpNetworkOperation(context: user.serializedDict())
        userSignUpOperation.perform(callBack: wrappedCallback)
    }
    
    func resetPassword(user: UserEntity, callback: @escaping Callback) {
        
        let wrappedCallback: Callback = { [weak self] (_ success: Bool, _ result: Any?, _ context: Any?, _ error: ErrorEntity?) -> Void in
            
            if (success) {
                
                self?.notifyObserversWithPasswordResetSuccess(email: user.email)
                
            } else {
                
                self?.notifyObserversWithPasswordResetFail(error: error)
            }
            
            callback(success, result, nil, error)
        }
        
        let userResetPasswordOperation: NetworkOperationProtocol = networkOperationFactory.userResetPasswordNetworkOperation(context: user.serializedDict())
        userResetPasswordOperation.perform(callBack: wrappedCallback)
    }
    
    func getUserProfile(user: UserEntity, callback: @escaping Callback) {
        
        let wrappedCallback: Callback = { [weak self] (_ success: Bool, _ result: Any?, _ context: Any?, _ error: ErrorEntity?) -> Void in
            
            if (success) {
                
                self?.user = result as? UserEntity
                
            } else {
                
                self?.notifyObserversWithPasswordResetFail(error:error)
            }
            
            callback(success, result, nil, error)
        }
        
        let userProfileOperation: NetworkOperationProtocol = networkOperationFactory.userProfileNetworkOperation(context: user.serializedDict())
        userProfileOperation.perform(callBack: wrappedCallback)
    }
    
    func logout(callback: @escaping Callback) {
        
        var error: ErrorEntity

        // [self.pushNotificationManager logoutUserWithCallback:^(BOOL success, id result, NSError *error) {
        // ...
        //    }];
        
        // Clear the token
        if let keychainError = self.keychainManager.setAuthentificationToken(token: nil) {
            
            error = ErrorEntity.init(code: 0, message: localizedString(key: "UserManager logout: unable to clear token in keychain. " + keychainError.message))
            
            self.notifyObserversWithLogoutFail(error: error)
            
            callback(false, nil, nil, error)

        } else {
            
            settingsManager.loggedInUser = nil
            
            let loggedOutUser: UserEntity = self.user!
            
            self.user = nil
            
            // There should be a webservice which does login and invalides token (if somebody sniffed and stole it). However there's no such webservice and we do logout locally only and so always consider logout as success immediatelly
            self.notifyObserversWithLogoutSuccess(user: loggedOutUser)
            
            callback(true, nil, nil, nil)
        }
    }
    
    // MARK: State observers

    func addServiceObserver(observer: UserManagerObserver, notificationType: UserManagerNotificationType, callback: @escaping Callback, context: Any?) {
        
        _observers.add(observer: observer as AnyObject, notificationName: notificationType.rawValue, callback: callback, context: context)
    }
    
    func removeServiceObserver(observer: UserManagerObserver, notificationType: UserManagerNotificationType) {
        
        _observers.remove(observer: observer as AnyObject, notificationName: notificationType.rawValue)
    }
    
    func removeServiceObserver(observer: UserManagerObserver) {
        
        _observers.remove(observer: observer as AnyObject)
    }
    
    // MARK:
    // MARK: Internal
    // MARK:
    
    // http://mikebuss.com/2014/06/22/lazy-initialization-swift/
    internal lazy var _observers: ObserverListProtocol = GMObserverList.init(observedSubject: self)
    //lazy var _user: UserEntity = UserEntity()
    //_user = [[UserObject alloc] init]; haven't ported this line. IMHO it's not needed anymore, maybe legacy stuff
    
    internal func save(user: UserEntity) -> ErrorEntity? {
        
        var result: ErrorEntity?
        
        if let token = user.token {
            
            // Store token
            if let error = self.keychainManager.setAuthentificationToken(token: token) {
                
                result = error
                
            } else {
                
                // Serialize and store user in settings
                let dict: Dictionary <String, AnyObject>? = user.serializedDict()
                settingsManager.loggedInUser = dict
                
                // Update in-memory user
                self.user = user
            }
            
        } else {
            
            result = ErrorEntity.init(code: 0, message: localizedString(key: "UserManager: unable to save user because token is empty"))
        }
        
        return result
    }
    
    // MARK: UserManagerObserver handling
    
    internal func notifyObserversWithLoginSuccess(user: UserEntity) {
        
        _observers.notifyObservers(notificationName: UserManagerNotificationType.didLogin.rawValue)
    }

    internal func notifyObserversWithLoginFail(error: ErrorEntity?) {
     
        _observers.notifyObservers(notificationName: UserManagerNotificationType.didFailedToLogin.rawValue)
    }
    
    internal func notifyObserversWithSignUpSuccess(user: UserEntity) {
        
        _observers.notifyObservers(notificationName: UserManagerNotificationType.didSignUp.rawValue)
    }
    
    internal func notifyObserversWithSignUpFail(error: ErrorEntity?) {
        
        _observers.notifyObservers(notificationName: UserManagerNotificationType.didFailedToSignUp.rawValue)
    }
    
    internal func notifyObserversWithUpdateSuccess(user: UserEntity) {
        
        _observers.notifyObservers(notificationName: UserManagerNotificationType.didUpdate.rawValue)
    }
    
    internal func notifyObserversWithUpdateFail(error: ErrorEntity?) {
        
        _observers.notifyObservers(notificationName: UserManagerNotificationType.didFailedToUpdate.rawValue)
    }
    
    internal func notifyObserversWithLogoutSuccess(user: UserEntity) {
    
        _observers.notifyObservers(notificationName: UserManagerNotificationType.didLogout.rawValue)
    }
    
    internal func notifyObserversWithLogoutFail(error: ErrorEntity?) {
        
        _observers.notifyObservers(notificationName: UserManagerNotificationType.didFailedToLogout.rawValue)
    }

    internal func notifyObserversWithPasswordResetSuccess(email: String) {
        
        _observers.notifyObservers(notificationName: UserManagerNotificationType.didResetPassword.rawValue)
    }
    
    internal func notifyObserversWithPasswordResetFail(error: ErrorEntity?) {
        
        _observers.notifyObservers(notificationName: UserManagerNotificationType.didFailedToResetPassword.rawValue)
    }
}
